src/Adapter/Product/Stock/QueryHandler/AbstractGetStockMovementHistoryHandler.php

```diff
@@ -0,0 +1,111 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\Product\Stock\QueryHandler;
+
+use PrestaShop\PrestaShop\Adapter\Product\Stock\Repository\StockAvailableMultiShopRepository;
+use PrestaShop\PrestaShop\Adapter\Product\Stock\Repository\StockMovementHistorySettings;
+use PrestaShop\PrestaShop\Adapter\Product\Stock\Repository\StockMovementRepository;
+use PrestaShop\PrestaShop\Core\Domain\Product\Stock\QueryResult\StockMovementHistory;
+use PrestaShop\PrestaShop\Core\Domain\Product\Stock\ValueObject\StockId;
+
+abstract class AbstractGetStockMovementHistoryHandler
+{
+    /**
+     * @var StockAvailableMultiShopRepository
+     */
+    protected $stockAvailableRepository;
+
+    /**
+     * @var StockMovementRepository
+     */
+    protected $stockMovementRepository;
+
+    public function __construct(
+        StockAvailableMultiShopRepository $stockAvailableRepository,
+        StockMovementRepository $stockMovementRepository
+    ) {
+        $this->stockAvailableRepository = $stockAvailableRepository;
+        $this->stockMovementRepository = $stockMovementRepository;
+    }
+
+    /**
+     * @return StockMovementHistory[]
+     */
+    protected function createStockMovementHistory(StockId $stockId, int $offset, int $limit): array
+    {
+        $historySettings = new StockMovementHistorySettings();
+        // Filter stock movements by stock ID
+        $historySettings->getMainFilter()->setStockIds($stockId);
+        // Select stock movements without order as single histories, excluded from groupings
+        $historySettings->getSingleFilter()->setIsOrder(false);
+        // Exclude stock movement groupings with zero quantity
+        $historySettings->excludeGroupingWithZeroQuantity(true);
```

Comment:
First time in my life. I read that implementing a complex use case with a clean and open design is not as good than hard coded. Modern development is all about open designs and anticipation, not about short term doodling.

`StockMovementHistorySettings` has been introduced only when the feature requirements have been changed after my 1st solution, to keep code clean and tidy. The new behaviour had nothing to do with filters, it relates to SQL grouping condition. Please consider this "extra" object as a convenient implementation of the pattern command :
https://refactoring.guru/design-patterns/command

I apologize for the extra effort to keep the behavior configurable and extendable.
I apologize for introducing a paginable behaviour, because it was easy and could help another developer in the future.
I apologize for using a modern and extendable design, and for providing a pattern for similar features in the future.
I apologize also for just doing my job.

Reply:
I agree, and very well at that! I get the feeling my suggestions were considered as "attacks", it wasn't the intent. My tone may have been a bit harsh but the intent is not to impose my ideas (which are no better than others), only to discuss and start a debate